Enhanced version of TicTacToe with tie handling, scoreboard and new game prompt

linncy opened the pull request:
### The version in this pull request will add three new features to TicTacToe.

**1.Handling the tied game**

- Add functionality to check the tie. A message will notify the player if the game is in a tie and will end the game.
- A new function (boardIsFull) is developed for checking the tie.

**2.Scoreboard functionality**

- Add a scoreboard functionality to keep track of how many games were won, lost, tied by each player.

- A new class (Scoreboard) is developed with two methods, score and printScoreBoard respectively.

**3.New game prompt**

- The game will not terminate right away whenever a player wins or the game ends up in a tie.
- Add functionality to prompt the user whether they want to play another game or not. If yes, a new game will be started. If not, the program will display the current scoreboard and then terminate.

ghaemisr committed:
Added feature1: Game has a tie situation

linncy committed:
Merge pull request #4 from ghaemisr/Feature1

Added feature1: Game has a tie situation

linncy committed:
Fix #2 Scoreboard

Add a new class Scoreboard for #2
Method:
1. void score(Player player) // Score up points after a game. Pass winnerr as the parameter
2. void printScoreBoard() // Print the current scoreboard

ghaemisr committed:
Merge pull request #3 from ghaemisr/Feature1

Added feature1: Game has a tie situation

ghaemisr committed:
Merge pull request #5 from ghaemisr/Feature2

Feature2 Add a Scoreboard

ghaemisr committed:
Assigned winner the value Player.NONE in the tie case

@linncy  To address the comment in issue #1, I assigned winner the value Player.NONE in the tie case. I changed the value in playMove function.

ghaemisr committed:
Merge pull request #6 from ghaemisr/Feature1

Assigned winner the value Player.NONE in the tie case

ghaemisr committed:
Added multiple play option. Waiting for ScoreBoard

To address issue #7 I added multiple game option and I am waiting for the score board bug to be fixed (issue #8).

linncy committed:
Fix bug #8 in ScoreBoard class

Fix bug #8 in ScoreBoard class
Use Board.Player instead of defining a new one.

linncy committed:
Merge pull request #9 from ghaemisr/Feature2

Fix bug #8 in ScoreBoard class

ghaemisr committed:
Merge pull request #10 from ghaemisr/master

Getting the fixed version after the bug

ghaemisr committed:
Added score board to multiple games

The score board is added to the multiple game version. #7 is finished.

ghaemisr committed:
Edited pom.xml and added showing scoreboard before terminating the game

ghaemisr committed:
Merge pull request #11 from ghaemisr/Feature3

Feature3: Finished feature 3 and edited the pom.xml file

linncy committed:
Specify java version in pom.xml

 Specify both the source and the target java version as 1.8

ghaemisr committed:
Merge pull request #12 from ghaemisr/Feature3

Specify java version in pom.xml

SaraGhlm commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java:
```diff
@@ -128,5 +130,13 @@ public Player getPlayerAtPos(int row, int col){
     }
 
-
+    private boolean boardIsFull(){
```
It is a good idea to have documentation for functions.

ghaemisr replied:
This is addressed in commit: abc14b9

SaraGhlm commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/ScoreBoard.java:
```diff
@@ -0,0 +1,48 @@
+
+public class ScoreBoard {
+    private int X_win,X_loss,X_tie;
+    private int O_win,O_loss,O_tie;
```
X_tie and O_tie are the same variables. You only need one tie variable.

linncy replied:
We fixed it in commit 815e331.
We removed the redundant variables "X_loss" and "O_loss", and replaced "X_tie" and "Y_tie" with one variable "tie".

alirezagoli commented on src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java:
```diff
@@ -31,7 +32,7 @@ public void promptNextPlayer(){
     public void playGame(){
         Scanner keyboardScanner = new Scanner(System.in);
-
+        board.initBoard();
```
It's better to create a new instance of the board class and use the constructor of the class to initialize all variables whenever you want to start a new game instead of calling initBoard() method. This way you can make sure that all variables are initialized correctly and also makes the extension of the class easier. For example, you forgot to initialize the current player at the end of each round.
```suggestion
        board= new Board();
```

ghaemisr replied:
I forgot to initialize the currentPlayer variable. Thanks for mentioning. 
This is addressed in commit: abc14b9

SaraGhlm reviewed (changes requested):
Please apply the following comments to your code.

alirezagoli commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java:
```diff
@@ -19,11 +19,11 @@ public Board(){
     }
 
-    private void initBoard(){
+    public void initBoard(){
```
It's better to keep this method private.
```suggestion
    private void initBoard(){
```

ghaemisr replied:
I changed this method to public because I used the function outside of the class.

alirezagoli commented on src/main/java/cs/ualberta/cmput402/tictactoe/board/ScoreBoard.java:
```diff
@@ -0,0 +1,48 @@
+package cs.ualberta.cmput402.tictactoe.board;
+
+public class ScoreBoard {
+    private int X_win,X_loss,X_tie;
```
Please use consistent naming through the project (either snake_case or Camelcase).

linncy replied:
We fixed it in commit 815e331.
We changed the variable names to follow the lower camel case naming convention.

alirezagoli reviewed (changes requested):
I reviewed your enhancements on TicTacToe, and I have some comments on your changes.  Please apply them to your code.

ghaemisr committed:
Applied three of the changes that was requested by the partner team

Changed the board initialization, added some comments to each new function and changed initBoard function to private.

linncy committed:
Update ScoreBoard.java to remove redundant varaibles and change variable names

Update ScoreBoard.java
1. Remove redundant varaibles "X_loss" and "O_loss", and replace "X_tie" and "Y_tie" with one variable "tie".
2. Change variable names to follow the lower camel case naming convention

 Fix issue #13
